source/agora/common/Types.d

```diff
@@ -73,12 +73,20 @@ public struct Height
         return Height(this.value++);
     }
 
-    /// Allow to offset an height by a fixed number
+    /// Allow to offset height by an addition of a fixed number
     public Height opBinary (string op : "+") (ulong offset) const
     {
         return Height(this.value + offset);
     }
 
+    /// Allow to offset height by a subtraction of a fixed number
+    public Height opBinary (string op : "-") (ulong offset) const
+    {
+        import std.conv : to;
+        assert(this.value - offset >= 0, "Attempt to subtract" ~ offset.to!string ~ " from height " ~ this.value.to!string);
+        return Height(this.value - offset);
+    }
```

Comment:
There two main issues with this code:
- It doesn't work. In D (and C, and C++, and probably many other languages), `unsigned - unsigned` gives you an unsigned result. Which means this condition will *never* be false, it will just wrap around. That's why all the underflow checks in Agora use `assert(offset >= value);`.
- It's using `assert`. Using `assert` in itself isn't such a problem, but the calling code would always have to make sure that `offset` will not trigger the value. If the end code ends up looking like this:
```D
if (offset >= value)
    offset = offset - value;
```
Then there's not much interest in this function, which is why I asked why you needed this.
In general, this approach works well if you deal with data you know has been validated, but I think that it's not really the case here.

Reply:
Another thing, you might want to look `text` (`std.conv` IIRC, maybe `std.format`):
```D
        assert(/* .... */, text("Attempt to subtract ", offset, " from height ", this.value);
```

Reply:
Yes I see what you mean. I'll drop that commit and re-push.

Reply:
Having removed it I see I am using it after all. I will push an update.

Reply:
Removed.